src/main/java/dev/blaauwendraad/masker/json/JsonPathState.java

```diff
@@ -0,0 +1,103 @@
+package dev.blaauwendraad.masker.json;
+
+import org.jspecify.annotations.Nullable;
+
+import java.util.ArrayDeque;
+
+class JsonPathState {
+    private static final KeyMatcher.StatefulRadixTrieNode NULL_NODE = new KeyMatcher.StatefulRadixTrieNode(new KeyMatcher.RadixTrieNode(new byte[0], new byte[0]), 0);
+
+    private final KeyMatcher keyMatcher;
+    private final ArrayDeque<KeyMatcher.StatefulRadixTrieNode> jsonPathSegments = new ArrayDeque<>();
+
+    JsonPathState(KeyMatcher keyMatcher) {
```

Comment:
I'd call this class "JsonPathTracker" as it also has a lookup logic besides storing the jsonpath state.